nsxt/resource_nsxt_manager_cluster.go

```diff
@@ -0,0 +1,323 @@
+/* Copyright © 2023 VMware, Inc. All Rights Reserved.
+   SPDX-License-Identifier: MPL-2.0 */
+
+package nsxt
+
+import (
+	"fmt"
+	"log"
+	"net"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/vmware/vsphere-automation-sdk-go/runtime/core"
+	"github.com/vmware/vsphere-automation-sdk-go/runtime/security"
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt-mp/nsx"
+	nsxModel "github.com/vmware/vsphere-automation-sdk-go/services/nsxt-mp/nsx/model"
+)
+
+func resourceNsxtManagerCluster() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceNsxtManagerClusterCreate,
+		Read:   resourceNsxtManagerClusterRead,
+		Update: resourceNsxtManagerClusterUpdate,
+		Delete: resourceNsxtManagerClusterDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"revision": getRevisionSchema(),
+			"cluster_id": {
+				Type:        schema.TypeString,
+				Description: "ID of the NSXT cluster",
+				Computed:    true,
+			},
+			"nodes": {
+				Type:        schema.TypeList,
+				Description: "Nodes in the cluster",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:        schema.TypeString,
+							Description: "ID of the cluster node",
+							Computed:    true,
+						},
+						"ip_address": {
+							Type:         schema.TypeString,
+							Description:  "IP addresses of the cluster node that will join the cluster of the host node",
+							ValidateFunc: validation.IsIPAddress,
+							Required:     true,
+						},
+						"password": {
+							Type:        schema.TypeString,
+							Description: "The authentication password for login",
+							Required:    true,
+							Sensitive:   true,
+						},
+						"username": {
+							Type:        schema.TypeString,
+							Description: "The username for login",
+							Required:    true,
+						},
+					},
+				},
+				Required: true,
+			},
+		},
+	}
+}
+
+type NsxClusterNode struct {
+	ID        *string
+	IPAddress *string
+	UserName  *string
+	Password  *string
+}
+
+func getClusterNodesFromSchema(d *schema.ResourceData) []NsxClusterNode {
+	nodes := d.Get("nodes").([]interface{})
+	var clusterNodes []NsxClusterNode
+	for _, node := range nodes {
+		data := node.(map[string]interface{})
+		id := data["id"].(string)
+		ipAddress := data["ip_address"].(string)
+		userName := data["username"].(string)
+		password := data["password"].(string)
+		nodeObj := NsxClusterNode{
+			ID:        &id,
+			IPAddress: &ipAddress,
+			UserName:  &userName,
+			Password:  &password,
+		}
+		clusterNodes = append(clusterNodes, nodeObj)
+	}
+	return clusterNodes
+}
+
+func setNodesInSchema(d *schema.ResourceData, nodes []NsxClusterNode) error {
+	nodeIPToCredentialMap := getIPtoCredentialMap(d)
```

Comment:
Could you please explain why is map needed here?

Reply:
The credential(username/password) is something we used when initialize a new client on for the new node. However in read function, this information is not in the clusterConfig that we obtained fromthe cluster client, so when we setNodesInSchema we need to a way to first retrieve the username/password from old schema, then set them in the elem, that's why we use this map to store such information
`elem["username"] = nodeIPToCredentialMap[*node.IPAddress][0]`
`elem["password"] = nodeIPToCredentialMap[*node.IPAddress][1]`

Reply:
Got it, thanks. I think we should add some computed info to those nodes, such as `status`, `fqdn`.